src/BlockTypes/Cart.php

```diff
@@ -65,33 +62,70 @@ public function render( $attributes = array(), $content = '' ) {
+		if ( did_action( 'woocommerce_blocks_cart_enqueue_data' ) ) {
+			return;
+		}
+
 		$data_registry = Package::container()->get(
-			\Automattic\WooCommerce\Blocks\Assets\AssetDataRegistry::class
+			AssetDataRegistry::class
 		);
-		if ( ! empty( $attributes['checkoutPageId'] ) && ! $data_registry->exists( 'page-' . $attributes['checkoutPageId'] ) ) {
-			$permalink = get_permalink( $attributes['checkoutPageId'] );
-			if ( $permalink ) {
-				$data_registry->add( 'page-' . $attributes['checkoutPageId'], $permalink );
+
+		$block_data = [
+			'shippingCountries' => [ WC()->countries, 'get_shipping_countries' ],
+			'shippingStates'    => [ WC()->countries, 'get_shipping_country_states' ],
+		];
+
+		foreach ( $block_data as $key => $callback ) {
+			if ( ! $data_registry->exists( $key ) ) {
+				$data_registry->add( $key, call_user_func( $callback ) );
 			}
 		}
-		if ( ! $data_registry->exists( 'shippingCountries' ) ) {
-			$data_registry->add( 'shippingCountries', WC()->countries->get_shipping_countries() );
-		}
-		if ( ! $data_registry->exists( 'shippingStates' ) ) {
-			$data_registry->add( 'shippingStates', WC()->countries->get_shipping_country_states() );
-		}
-		if ( ! $data_registry->exists( 'cartData' ) ) {
-			$data_registry->add( 'cartData', WC()->api->get_endpoint_data( '/wc/store/cart' ) );
+
+		$permalink = ! empty( $attributes['checkoutPageId'] ) ? get_permalink( $attributes['checkoutPageId'] ) : false;
+
+		if ( $permalink && ! $data_registry->exists( 'page-' . $attributes['checkoutPageId'] ) ) {
+			$data_registry->add( 'page-' . $attributes['checkoutPageId'], $permalink );
 		}
+
 		if ( ! $data_registry->exists( 'quantitySelectLimit' ) ) {
 			/**
 			 * Note: this filter will be deprecated if/when quantity select limits
 			 * are added at the product level.
 			 *
 			 * @return {integer} $max_quantity_limit Maximum quantity of products that can be selected in the cart.
 			 */
-			$max_quantity_limit = apply_filters( 'woocommerce_maximum_quantity_selected_cart', 99 );
-			$data_registry->add( 'quantitySelectLimit', $max_quantity_limit );
+			$data_registry->add( 'quantitySelectLimit', apply_filters( 'woocommerce_maximum_quantity_selected_cart', 99 ) );
+		}
+
+		// Hydrate the following data depending on admin or frontend context.
+		if ( ! is_admin() ) {
+			$this->hydrate_from_api( $data_registry );
+		}
+
+		do_action( 'woocommerce_blocks_cart_enqueue_data' );
+	}
+
+	/**
+	 * Register/enqueue scripts used for this block.
+	 *
+	 * @param array $attributes  Any attributes that currently are available from the block.
+	 *                           Note, this will be empty in the editor context when the block is
+	 *                           not in the post content on editor load.
+	 */
+	protected function enqueue_scripts( array $attributes = [] ) {
+		Assets::register_block_script( $this->block_name . '-frontend', $this->block_name . '-block-frontend' );
+
+		do_action( 'woocommerce_blocks_cart_enqueue_scripts' );
```

Comment:
We'll need to rename this back to `woocommerce_blocks_enqueue_cart_block_scripts_after` because there is tracking that was hooked into it I think.